src/clip.sol

```diff
@@ -0,0 +1,472 @@
+    mapping (address => uint256) public wards;
+    function rely(address usr) external auth { wards[usr] = 1; emit Rely(usr); }
+    function deny(address usr) external auth { wards[usr] = 0; emit Deny(usr); }
+    modifier auth {
+        require(wards[msg.sender] == 1, "Clipper/not-authorized");
+        _;
+    }
+
+    // --- Data ---
+    bytes32  immutable public ilk;   // Collateral type of this Clipper
+    VatLike  immutable public vat;   // Core CDP Engine
+
+    DogLike     public dog;      // Liquidation module
+    address     public vow;      // Recipient of dai raised in auctions
+    SpotterLike public spotter;  // Collateral price module
+    AbacusLike  public calc;     // Current price calculator
+
+    uint256 public buf;    // Multiplicative factor to increase starting price                  [ray]
+    uint256 public tail;   // Time elapsed before auction reset                                 [seconds]
+    uint256 public cusp;   // Percentage drop before auction reset                              [ray]
+    uint64  public chip;   // Percentage of tab to suck from vow to incentivize keepers         [wad]
+    uint192 public tip;    // Flat fee to suck from vow to incentivize keepers                  [rad]
+    uint256 public chost;  // Cache the ilk dust times the ilk chop to prevent excessive SLOADs [rad]
+
+    uint256   public kicks;   // Total auctions
+    uint256[] public active;  // Array of active auction ids
+
+    struct Sale {
+        uint256 pos;  // Index in active array
+        uint256 tab;  // Dai to raise       [rad]
+        uint256 lot;  // collateral to sell [wad]
+        address usr;  // Liquidated CDP
+        uint96  tic;  // Auction start time
+        uint256 top;  // Starting price     [ray]
+    }
+    mapping(uint256 => Sale) public sales;
+
+    uint256 internal locked;
+
+    // Levels for circuit breaker
+    // 0: no breaker
+    // 1: no new kick()
+    // 2: no new kick() or redo()
+    // 3: no new kick(), redo(), or take()
+    uint256 public stopped = 0;
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+
+    event File(bytes32 indexed what, uint256 data);
+    event File(bytes32 indexed what, address data);
+
+    event Kick(
+        uint256 indexed id,
+        uint256 top,
+        uint256 tab,
+        uint256 lot,
+        address indexed usr,
+        address indexed kpr,
+        uint256 coin
+    );
+    event Take(
+        uint256 indexed id,
+        uint256 max,
+        uint256 price,
+        uint256 owe,
+        uint256 tab,
+        uint256 lot,
+        address indexed usr
+    );
+    event Redo(
+        uint256 indexed id,
+        uint256 top,
+        uint256 tab,
+        uint256 lot,
+        address indexed usr,
+        address indexed kpr,
+        uint256 coin
+    );
+
+    event Yank(uint256 id);
+
+    // --- Init ---
+    constructor(address vat_, address spotter_, address dog_, bytes32 ilk_) public {
```

Comment:
Constructor visibility is redundant. It is assumed to be public and can be safely removed.